packages/plugin-layout/src/LayoutEngine.ts

```diff
@@ -0,0 +1,277 @@
+import { Constructor } from '../../utils/src/utils';
+import { VNode } from '../../core/src/VNodes/VNode';
+import JWEditor from '../../core/src/JWEditor';
+import { ZoneIdentifier, ZoneNode } from './ZoneNode';
+import { FragmentNode } from '../../core/src/VNodes/FragmentNode';
+
+export type componentReference = string | VNode;
+export type DomZonePosition = 'after' | 'prepend' | 'append' | 'replace';
+
+export interface Component<N extends Constructor<VNode> = Constructor<VNode>> {
+    zone: ZoneIdentifier;
+    VNode?: N | InstanceType<N>;
+    args?: ConstructorParameters<N>;
+    id?: componentReference;
+    defaultHidden?: boolean;
+}
+export interface VTemplate<T = {}> {
+    node?: VNode;
+    target: T;
+    position: DomZonePosition;
+}
+
+export type LayoutEngineConfig<C extends VTemplate = VTemplate> = {
+    templates?: C[];
+};
+
+export class LayoutEngine<
+    T extends {} = {},
+    V extends VTemplate<T> = VTemplate<T>,
+    C extends Component = Component
+> {
+    protected templates: V[];
+
+    constructor(public editor: JWEditor, public configuration: LayoutEngineConfig<V> = {}) {}
+
+    protected Components: C[] = [];
+    protected components: Map<componentReference, VNode[]> = new Map();
+    protected references: Map<VNode, componentReference> = new Map();
+
+    //--------------------------------------------------------------------------
+    // Lifecycle
+    //--------------------------------------------------------------------------
+
+    /**
+     * Load the given Component in this layout engine.
+     *
+     * @param Component
+     */
+    load(Component: C): void {
+        this.Components.push(Component);
+    }
+    async start(): Promise<void> {
+        this.templates = [];
+        const zoneNodes: ZoneNode[] = [];
+        for (const template of this.configuration.templates) {
+            this.templates.push(template);
+            zoneNodes.push(...template.node.descendants(ZoneNode));
```

Comment:
What if `template.node` is empty as the node could be not set as written in the interface?

Reply:
It is indeed impossible. there are always vNodes, even if the user gives a template because it is parsed.